src/semaphore.c

```diff
@@ -804,3 +880,130 @@ _dispatch_thread_semaphore_wait(_dispatch_thread_semaphore_t sema)
 #error "No supported semaphore type"
 #endif
 }
+
+#if USE_FUTEX_SEM
+#define DISPATCH_FUTEX_NUM_SPINS 100
+#define DISPATCH_FUTEX_VALUE_MAX INT_MAX
+#define DISPATCH_FUTEX_NWAITERS_SHIFT 32
+#define DISPATCH_FUTEX_VALUE_MASK ((1ull << DISPATCH_FUTEX_NWAITERS_SHIFT) - 1)
+
+static int _dispatch_futex_trywait(dispatch_futex_t *dfx);
+static int _dispatch_futex_wait_slow(dispatch_futex_t *dfx,
+		const struct timespec* timeout);
+static int _dispatch_futex_syscall(dispatch_futex_t *dfx, int op, int val,
+		const struct timespec *timeout);
+
+/* all memory orders below were added with strict seq_cst semantics
+ * assumed to be the strongest
+ */
+
+int
+_dispatch_futex_init(dispatch_futex_t *dfx)
+{
+	dfx->dfx_data = 0;
+	return 0;
+}
+
+int
+_dispatch_futex_dispose(dispatch_futex_t *dfx)
+{
+	(void)dfx;
+	return 0;
+}
+
+// Increments semaphore value and reads wait count in a single atomic
+// operation, with release MO. If wait count is nonzero, issues a FUTEX_WAKE.
+int
+_dispatch_futex_signal(dispatch_futex_t *dfx)
+{
+	uint64_t orig;
+	do {
+		orig = dfx->dfx_data;
+		if (slowpath((orig & DISPATCH_FUTEX_VALUE_MASK) ==
+					 DISPATCH_FUTEX_VALUE_MAX)) {
+			DISPATCH_CRASH("semaphore overflow");
+		}
+	} while (!dispatch_atomic_cmpxchg2o(dfx, dfx_data, orig, orig + 1, seq_cst));
```

Comment:
yeah the definition is right here:

```C
#define dispatch_atomic_cmpxchgvw(p, e, v, g, m) \
		({ _dispatch_atomic_basetypeof(p) _v = (v), _r = (e); _Bool _b = \
		__c11_atomic_compare_exchange_weak(_dispatch_atomic_c11_atomic(p), \
		&_r, _v, dispatch_atomic_memory_order_##m, \
		dispatch_atomic_memory_order_relaxed); *(g) = (typeof(*(p)))_r;  _b; })
```

the 2o variants are for "reasons" but just about `&p->field` where `p` and `field` are the two first arguments.

that thing above looks weird but does something akin to this (pseudo code):

```C
#define dispatch_atomic_cmpxchgvw(p, e, v, g, m) \
		({ *g = e;  \
		atomic_compare_exchange_weak_explicit(p, g, v, memory_order_##m, memory_order_relaxed); })
```

the "v" variant does the same for strong.
and the non "v" variant is a strong cmpxchg that doesn't get you the current value.

rule of thumb, if you loop, you usually want a `v` variant as it will reload the value for you efficiently, and usually you want the `w` (weak) variant as going to strong on LL/SC architectures such as arm or ppc will itself loop, so if you do, you should piggy back on your own loop instead.